### ISSUE TITLE
jax.lax.pswapaxes calculates incorrect shapes when using soft_pmap

### ISSUE THREAD

**Issue by Thenerdstation** [google/jax#1672]
Code:
```python
def f(a):
  a = jax.lax.pswapaxes(a, "i", 0)
  print(a.shape)

print("pmap:")
jax.pmap(f, "i")(onp.ones((8, 8)))
print("\nsoft_pmap:")
jax.soft_pmap(f, "i")(onp.ones((8, 8)))
```

Results:
```
pmap:
(8,)

soft_pmap:
(1,)
```

**Comment by Thenerdstation** [comment]
Note: This only happens when using a TPU.

When testing this with a CPU, results seem ok:
```
pmap:
(8,)

soft_pmap:
(8,)
```

**Comment by levskaya** [comment]
1) the above code snippet is a bit weird, jax transforms operate on functions via data-dependent tracing (purely functional inputs -> outputs), so the above is a gonzo function to hand jax to begin with, let's look at:
```python
def f(a):
  return jax.lax.pswapaxes(a, "i", 0)

print("pmap:")
print(jax.pmap(f, "i")(onp.ones((8, 8))).shape)
print("\nsoft_pmap:")
print(jax.soft_pmap(f, "i")(onp.ones((8, 8))).shape)
```
this still yields a difference:
```
pmap:
(8, 8)

soft_pmap:
(64, 1)
```

2) This is happening because `pswapaxes` is defined in terms of `all_to_all` which currently has incorrect `abstract_eval` rule for the underlying xla all_to_all function and and an incorrect `split_axis` rule (which soft_pmap uses). (bug: #1332)  We're trying to figure out the fix in PR #1577 alongside the right way to expose all-to-all in jax given it's hidden/named dimension convention in pmap.

3) I'm curious - how are you using pswapaxes?  We're curious what people actually want to do with pswapaxes/all_to_all to help guide that discussion.

**Comment by Thenerdstation** [comment]
Specifically, we're trying to do [tensor network](https://github.com/google/TensorNetwork) algorithms on TPUs, which is basically doing a ton of various `np.tensordot` operations.

To allow us to run across multiple TPU cores, we need to "shard" an axis of a tensor so that the tensor is split up over the cores via `pmap`. However, sometimes we'll need to do operations on this sharded axis, so we'll need to use `pswapaxes` to make that "sharded" axis "local" and make a different "local" axis "sharded". After we do the `pswapaxes`, we can freely do operations on that axis.

Having a correct `soft_pmap` would be really nice for us, as it would allow us to write code that could theoretically work regardless of how large of a TPU slice we actually use.
